KAFKA-1981 Make log compaction point configurable

ewasserman opened the pull request:
Reduced scope to controlling only minimum time before compaction. Changed to using message time when available.

Commit:
remove size and message count constraints

Commit:
remove size and message count constraints

ijuma commented on core/src/main/scala/kafka/log/LogCleanerManager.scala:
```diff
@@ -240,3 +236,66 @@ private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[To
+    val firstDirtyOffset = {
+      val offset = lastCleanOffset.getOrElse(logStartOffset)
+      if (offset < logStartOffset) {
+        error("Resetting first dirty offset to log start offset %d since the checkpointed offset %d is invalid."
```
May as well change this to string interpolation like the other log statements.

junrao commented on core/src/main/scala/kafka/log/LogCleanerManager.scala:
```diff
@@ -240,3 +236,66 @@ private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[To
+        if (compactionLagMs > 0) {
+          dirtySegments.find {
+            case segment =>
+              segment.log.lastOption match {
```
lastOption will require a full scan of the log and won't be efficient. Also, the largest timestamp may not always be in the last message. If you wait for KAFKA-3163 (KIP-33) to be committed, you will get a more efficient and accurate api LogSegment.largestTimestamp.

ewasserman replied:
Do we still expect KAFKA-3163 to be committed for 0.10.1.0? I have no objection to using LogSegment.largestTimestamp if that's available.

junrao replied:
Yes, we do expect KAFKA-3163 to be committed for 0.10.1.0.

ewasserman replied:
Since KAFKA-3163 is still not merged can we go back to using the (less accurate but efficient) log segment modification time? Once the time indices are available compaction could be modified to use them in a subsequent PR.

junrao replied:
Let me ping the contributor of KAFKA-3163. I think it's pretty close to be merged.

junrao commented on core/src/main/scala/kafka/log/LogCleanerManager.scala:
```diff
@@ -240,3 +236,66 @@ private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[To
+    * Returns the range of dirty offsets that can be cleaned.
+    *
+    * @param log the log
+    * @param lastClean the map of checkpointed offsets
```
We need to add topicAndPartition in the javadoc.

ijuma committed:
MINOR: Remove redundant clause in secureAclsEnabled check

Also include a few minor clean-ups.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Manikumar reddy O <manikumar.reddy@gmail.com>, Grant Henke <granthenke@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1623 from ijuma/fix-zk-inconsistent-security-check

ijuma committed:
KAFKA-3954; Consumer should use internal topics information returned by the broker

It previously hardcoded it.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Grant Henke <granthenke@gmail.com>, Jason Gustafson <jason@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1613 from ijuma/kafka-3954-consumer-internal-topics-from-broker

Commit:
MINOR: Make kafka-run-class.sh work under Cygwin

kafka-run-class.sh now runs under Cygwin; paths and classpath are set up properly.

**WARNING:**  The script was not tested on a Linux machine, only under Cygwin.  Prior to merge it into trunk, if accepted, please run a quick test to ensure nothing broke.  From my own code review, there should not be any problem, but we can never be too sure.

I do not have the environment to test it under Linux at this moment.

Author: deragon <32nx9812masakjds>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1487 from deragon/trunk

granthenke committed:
KAFKA-3934: Start scripts enable GC by default with no way to disable

…o disable

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1631 from granthenke/garbage-flag

Commit:
MINOR: Doc individual partition must fit on the server that host it

Author: sahil kharb <sahilkharb@iitj.ac.in>

Reviewers: Ismael Juma, Guozhang Wang

Closes #1709 from bitfurry/trunk

dguy committed:
KAFKA-3936: Validate parameters as early as possible

Added non null checks to parameters supplied via the DSL and `TopologyBuilder`

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Edward Ribeiro <edward.ribeiro@gmail.com>, Guozhang Wang <wangguoz@gmail.com>

Closes #1711 from dguy/kafka-3936

dguy committed:
KAFKA-3914: Global discovery of state stores

guozhangwang enothereska mjsax miguno  please take a look. A few things that need to be clarified

1. I've added StreamsConfig.USER_ENDPOINT_CONFIG, but should we have separate configs for host and port or is this one config ok?
2. `HostState` in the KIP has a byte[] field - not sure why and what it would be populated with
3. I've changed the API to return `List<KafkaStreamsInstance>` as opposed to `Map<HostInfo, Set<TaskMetadata>>` as i find this far more intuitive to work with.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Matthias J. Sax, Michael G. Noll, Eno Thereska, Guozhang Wang

Closes #1576 from dguy/kafka-3914v2

Commit:
KAFKA-4031; Check if buffer cleaner is null before using it

A small fix to check null before using the reference

Author: Som Sahu <sosahu@microsoft.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1718 from soumyajit-sahu/nullCheckForDirectBufferCleaner

ewencp committed:
KAFKA-3847: Use a separate producer per source task

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Jason Gustafson, Gwen Shapira

Closes #1727 from ewencp/kafka-3847-per-task-producers and squashes the following commits:

7d39724 [Ewen Cheslack-Postava] Add timeout for closing producers.
98ec7f6 [Ewen Cheslack-Postava] KAFKA-3847: Use a separate producer per source task

granthenke committed:
KAFKA-2946; DeleteTopic - protocol and server side implementation

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <me@ewencp.org>, Jun Rao <junrao@gmail.com>

Closes #1616 from granthenke/delete-wire-new

bbaugher committed:
KAFKA-4012; Added #toString() to KerberosShortNamer

Author: Bryan Baugher <bryan.baugher@cerner.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1694 from bbaugher/KAFKA-4012

resetius committed:
KAFKA-3997; log partition name on truncation

Author: Alexey Ozeritsky <aozeritsky@yandex-team.ru>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1724 from resetius/KAFKA-3997

hachikuji committed:
KAFKA-4034; Avoid unnecessary consumer coordinator lookup

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1720 from hachikuji/KAFKA-4034

granthenke committed:
KAFKA-2946; MINOR: Follow up. Add Delete to AclCommandTest

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jun Rao <junrao@gmail.com>

Closes #1730 from granthenke/test-delete

omkreddy committed:
KAFKA-4035; AclCommand should allow Describe operation on groups

Author: Manikumar Reddy O <manikumar.reddy@gmail.com>

Reviewers: Sriharsha Chintalapani <harsha@hortonworks.com>, Jun Rao <junrao@gmail.com>

Closes #1723 from omkreddy/KAFKA-4035

guozhangwang committed:
KAFKA-3769: Create new sensors per-thread in KafkaStreams

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Damian Guy <damian.guy@gmail.com>, Matthias J. Sax <matthias@confluent.io>, Michael G. Noll <michael@confluent.io>, Greg Fodor <gfodor@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1530 from guozhangwang/K3769-per-thread-metrics

enothereska committed:
HOTFIX: Re-inserted SimpleBenchmark output for system tests

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1744 from enothereska/hotfix-ducktape-marker

omkreddy committed:
KAFKA-4044; log actual socket send/receive buffer size after connecting in Selector

Author: Manikumar Reddy O <manikumar.reddy@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1750 from omkreddy/KAFKA-4044-LOG

hachikuji committed:
KAFKA-3888: send consumer heartbeats from a background thread (KIP-62)

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>, Guozhang Wang <wangguoz@gmail.com>

Closes #1627 from hachikuji/KAFKA-3888

granthenke committed:
KAFKA-4038; Transient failure in DeleteTopicsRequestTest.testErrorDeleteTopicRequests

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1737 from granthenke/transient-delete

ewencp committed:
KAFKA-4037: Make Connect REST API retries aware of 409 CONFLICT errors

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1733 from ewencp/rest-api-retries

vahidhashemian committed:
MINOR: Clarification in producer config documentation

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Dana Powers, Gwen Shapira

Closes #1732 from vahidhashemian/minor/clarify_producer_config_documentation

p-thorpe committed:
MINOR: Remove # from .bat start script

On Windows, the following output is seen when starting Zookeeper and Kafka servers:
```
'#' is not recognized as an internal or external command,
operable program or batch file.
```

This pull request makes a minor correction to the Windows `kafka-run-class.bat` script to replace the use of `#` with `rem`.

Author: P. Thorpe <paul.thorpe@uk.ibm.com>

Reviewers: Vahid Hashemian, Gwen Shapira

Closes #1740 from p-thorpe/trunk

ewencp committed:
KAFKA-3845: KIP-75: Add per-connector converters

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Shikhar Bhushan, Gwen Shapira

Closes #1721 from ewencp/kafka-3845-per-connector-converters

becketqin committed:
KAFKA-3163; Add time based index to Kafka.

This patch is for KIP-33.
https://cwiki.apache.org/confluence/display/KAFKA/KIP-33+-+Add+a+time+based+log+index

Author: Jiangjie Qin <becket.qin@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jun Rao <junrao@gmail.com>, Liquan Pei <liquanpei@gmail.com>

Closes #1215 from becketqin/KAFKA-3163

junrao commented:
@ewasserman : KAFKA-3163 is now committed. Could you rebase?

Commit:
KAFKA-4050; Allow configuration of the PRNG used for SSL

Add an optional configuration for the SecureRandom PRNG implementation, with the default behavior being the same (use the default implementation in the JDK/JRE).

Author: Todd Palino <Todd Palino>

Reviewers: Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Joel Koshy <jjkoshy@gmail.com>, Jiangjie Qin <becket.qin@gmail.com>, Rajini Sivaram <rajinisivaram@googlemail.com>

Closes #1747 from toddpalino/trunk

mimaison committed:
KAFKA-4056: Kafka logs values of sensitive configs like passwords

In case of unknown configs, only list the name without the value

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Jaikiran, Gwen Shapira, Grant Henke, Ryan Pridgeon, Dustin Cote

Closes #1759 from mimaison/KAFKA-4056

alnzng committed:
KAFKA-4053: remove redundant if/else statements in TopicCommand

Author: Shuai Zhang <shuai.xyz@gmail.com>

Reviewers: Gwen Shapira

Closes #1751 from sh-z/KAFKA-4053

enothereska committed:
KAFKA-4016: Added join benchmarks

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ismael Juma, Damian Guy, Guozhang Wang

Closes #1700 from enothereska/join-benchmarks

hachikuji committed:
KAFKA-3949: Fix race condition when metadata update arrives during rebalance

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Vahid Hashemian, Guozhang Wang

Closes #1762 from hachikuji/KAFKA-3949

becketqin committed:
KAFKA-3163; Minor follow up for (KIP-33)

junrao Could you take a look when get a chance? Thanks.

Author: Jiangjie Qin <becket.qin@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1769 from becketqin/KAFKA-3163-follow-up

tom-pang committed:
KAFKA-3894; log cleaner can partially clean a segment

As discussed in https://issues.apache.org/jira/browse/KAFKA-3894, this PR makes the log cleaner do a "partial" clean on a segment, whereby it builds a partial offset map up to a particular offset in a segment. Once cleaning resumes again, we will continue from the next dirty offset, which can now be located in the middle of a segment.

Prior to this PR, segments with overly numerous keys could crash the log cleaner thread, as it was required that the log cleaner had to fit at least a single segment in the offset map.

Author: Tom Crayford <tcrayford@googlemail.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1725 from tcrayford/dont_crash_log_cleaner_thread_if_segment_overflows_buffer

rajinisivaram committed:
KAFKA-4066; Fix NPE in consumer due to multi-threaded updates

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1763 from rajinisivaram/KAFKA-4066

ssaamm committed:
MINOR: Fix typos in security section

1. I think the instructions in step 2 of the security section which describe adding the CA to server/client truststores are swapped. That is, the instruction that says to add the CA to the server truststore adds it to the client truststore (and vice versa).
2. "clients keys" should be possessive ("clients' keys").

This contribution is my original work, and I license the work to the project under the project's open source license.

Author: Samuel Taylor <staylor@square-root.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1651 from ssaamm/trunk

mjsax committed:
MINOR: improve Streams application reset tool to make sure application is down

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1764 from mjsax/improveResetTool

guozhangwang committed:
MINOR: add slf4jlog4j to streams example

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1731 from guozhangwang/Kminor-log4j-streams-examples

rajinisivaram committed:
KAFKA-4051: Use nanosecond clock for timers in broker

Use System.nanoseconds instead of System.currentTimeMillis in broker timer tasks to cope with changes to wall-clock time.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Gwen Shapira

Closes #1768 from rajinisivaram/KAFKA-4051

guozhangwang committed:
KAFKA-4049: Fix transient failure in RegexSourceIntegrationTest

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <me@ewencp.org>

Closes #1746 from guozhangwang/K4049-RegexSourceIntegrationTest-failure

granthenke committed:
KAFKA-4032; Uncaught exceptions when autocreating topics

handled by adding a catch all for any unhandled exception. Because the jira specifically mentions the InvalidReplicationFactor exception, a test was added for that specific case.

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1739 from granthenke/create-errors

hachikuji committed:
KAFKA-3916; Check for disconnects properly before sending from the controller

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1734 from hachikuji/KAFKA-3916

guozhangwang committed:
MINOR: Refactor TopologyBuilder with ApplicationID Prefix

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Damian Guy <damian.guy@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1736 from guozhangwang/Kminor-topology-applicationID

ijuma committed:
KAFKA-4073; MirrorMaker should handle messages without timestamp correctly

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1773 from ijuma/kafka-4073-mirror-maker-timestamps

rajinisivaram committed:
KAFKA-3680; Enable Kafka clients to run in any classloader env

Configure default classes using class objects instead of class names, enable configurable lists of classes to be specified as class objects, add tests for different classloader configurations.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Sriharsha Chintalapani <harsha@hortonworks.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1421 from rajinisivaram/KAFKA-3680

kkonstantine committed:
KAFKA-2894: WorkerSinkTask should rewind offsets on rebalance

Author: Konstantine Karantasis <k.karantasis@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1771 from kkonstantine/KAFKA-2894-rewind-offsets-on-rebalance

Commit:
KAFKA-3937; Kafka Clients Leak Native Memory For Longer Than Needed With Compressed Messages

ijuma - Making the change against trunk based on your suggestions to have the stream closing handled in the private RecordIterator constructor which I understand is only to be used only if the block of message(s) are compressed.

Author: William Yu <wyu@unified.com>
Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1760 from wiyu/compressor_memory_leak_in_fetcher

ijuma committed:
KAFKA-4082; Upgrade to Gradle 3.0

Also upgrade scoverage (required for compatibility) and remove usage of
`useAnt` which doesn't exist in Gradle 3.0

It turns out that one cannot even run `gradle` to download the project Gradle version if `useAnt` is used in the build. This is unfortunate (the SBT launcher has much saner behaviour).

Release notes: https://docs.gradle.org/3.0/release-notes

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Grant Henke <granthenke@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>, Sriharsha Chintalapani <harsha@hortonworks.com>

Closes #1774 from ijuma/kafka-4082-support-gradle-3.0

repl-mathieu-fenniak committed:
MINOR: Update rocksDB dependency to 4.9.0

rocksdbjni version 4.9.0 now includes support for running on Windows; this PR updates Kafka Stream's dependency to that version.  Tests pass locally, except for a timeout in testReprocessingFromScratchAfterReset that doesn't seem related; it happens with and without this change.

This contribution is my original work and I license the work to the project under the project's open source license.

Author: Mathieu Fenniak <mathieu.fenniak@replicon.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1783 from mfenniak/update-rocksdb-4.9

sven0726 committed:
MINOR: Fix wording in Metadata.add javadoc

Author: sven0726 <sven0726@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1719 from sven0726/trunk

dguy committed:
KAFKA-4015; Change cleanup.policy config to accept a list of valid policies

Change cleanup.policy to accept a comma separated list of valid policies.
Updated LogCleaner.CleanerThread to also run deletion for any topics configured with compact,delete.
Ensure Log.deleteSegments only runs when delete is enabled.
Additional Integration and unit tests to cover new option

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Jun Rao <junrao@gmail.com>

Closes #1742 from dguy/kafka-4015

ewasserman committed:
fixes

ewasserman committed:
fixes
